polynote-kernel/src/main/scala/polynote/kernel/package.scala

```diff
@@ -51,31 +52,17 @@ package object kernel {
       case notEnd         => Stream.emit(Some(notEnd))
     }.unNoneTerminate
 
-  }
+    def terminateAfterEquals(value: A): Stream[RIO[R, ?], A] = terminateAfter(_ == value)
+
+    def interruptAndIgnoreWhen(signal: Promise[Throwable, Unit])(implicit concurrent: Concurrent[RIO[R, ?]]): Stream[RIO[R, ?], A] =
+      stream.interruptWhen(signal.await.either.as(Right(()): Either[Throwable, Unit]))
 
-  // some tuple syntax that ZIO doesn't natively have
-  // PR for including this in ZIO: https://github.com/zio/zio/pull/1444
-  final implicit class ZIOTuple4[E, RA, A, RB, B, RC, C, RD, D](
-    val zios4: (ZIO[RA, E, A], ZIO[RB, E, B], ZIO[RC, E, C], ZIO[RD, E, D])
-  ) extends AnyVal {
-    def map4[F](f: (A, B, C, D) => F): ZIO[RA with RB with RC with RD, E, F] =
-      for {
-        a <- zios4._1
-        b <- zios4._2
-        c <- zios4._3
-        d <- zios4._4
-      } yield f(a, b, c, d)
   }
 
-  final implicit class ZIOTuple3[E, RA, A, RB, B, RC, C](
-    val zios3: (ZIO[RA, E, A], ZIO[RB, E, B], ZIO[RC, E, C])
-  ) extends AnyVal {
-    def map3[F](f: (A, B, C) => F): ZIO[RA with RB with RC, E, F] =
-      for {
-        a <- zios3._1
-        b <- zios3._2
-        c <- zios3._3
-      } yield f(a, b, c)
+  final implicit class StreamUIOps[A](val stream: Stream[UIO, A]) {
+
+    def interruptAndIgnoreWhen(signal: Promise[Throwable, Unit])(implicit concurrent: Concurrent[UIO]): Stream[UIO, A] =
```

Comment:
Explanation of this:

I noticed that there are a lot of places where we do `stream.intteruptWhen(...)` and pass it a promise or something so that the stream will get interrupted when the promise completes. But it turns out that if that promise is failed with an error, that error makes it all the way up to the Platform which freaks out and logs it with a strongly-worded letter. Which makes perfect sense – those streams are unsupervised, and of course if you're passing them a signal of `Either[Throwable, Unit]` you should expect that they're going to do some kind of failure when they get a `Left`.

Anyway, usually we don't want to cause a failure in the stream when that promise fails – that failure is treated elsewhere. We just want to also interrupt the stream. So this extension method is for ignoring any failure in the interrupt signal. I replaced it in several places (in some spots I couldn't get IntelliJ to be happy with it so I just kinda duplicated this. Sorry)

Reply:
Nice! Looks like it cleaned up a few places 😄